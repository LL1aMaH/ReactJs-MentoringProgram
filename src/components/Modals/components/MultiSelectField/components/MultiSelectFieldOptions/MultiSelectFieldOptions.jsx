import React, { memo, useState, useCallback } from 'react';
import classnames from 'classnames';

import styles from '../../../../../../../styles/MultiSelectFieldOptions.module.scss';

import { MultiSelectFieldItem } from '../MultiSelectFieldItem/MultiSelectFieldItem';


export const MultiSelectFieldOptions = memo(function MultiSelectFieldOptionsBase(
  props,
) {
  const { valueMap, options, onChange, renderDescription, disabled = false } = props;

  const [activeOption, setActiveOption] = useState<string | null>(null);

  const handleMouseEnter = useCallback((option) => {
    setActiveOption(option);
  }, []);

  return (
    <div className={classnames(styles.layout)}>
      <ul className={styles.options}>
        {options.map((option) => {
          const optionLabel = option;

          return (
            <MultiSelectFieldItem
              key={optionLabel}
              value={optionLabel}
              label={optionLabel}
              checked={valueMap.includes(optionLabel)}
              disabled={disabled}
              option={option}
              onChange={onChange}
              onMouseEnter={handleMouseEnter}
            />
          );
        })}
      </ul>
      {renderDescription && activeOption && (
        <div className={styles.description}>{renderDescription(activeOption)}</div>
      )}
    </div>
  );
});
