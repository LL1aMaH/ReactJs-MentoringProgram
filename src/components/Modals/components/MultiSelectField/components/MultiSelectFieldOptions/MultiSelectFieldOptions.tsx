import React, { memo, useState, useCallback, ReactNode } from 'react';
import classnames from 'classnames';

import { MultiSelectFieldItem } from '../MultiSelectFieldItem';
import styles from './MultiSelectFieldOptions.css';

export type TMultiSelectFieldOptionsProps = {
  valueMap: Array<string>;
  options: Array<string>;
  onChange: (value: string) => void;
  renderDescription?: (option: string) => ReactNode;
  disabled?: boolean;
};

export const MultiSelectFieldOptions = memo(function MultiSelectFieldOptionsBase(
  props: TMultiSelectFieldOptionsProps,
): JSX.Element {
  const { valueMap, options, onChange, renderDescription, disabled = false } = props;

  const [activeOption, setActiveOption] = useState<string | null>(null);

  const handleMouseEnter = useCallback((option) => {
    setActiveOption(option);
  }, []);

  return (
    <div className={classnames(styles.layout)}>
      <ul className={styles.options}>
        {options.map((option) => {
          const optionLabel = option;

          return (
            <MultiSelectFieldItem
              key={optionLabel}
              value={optionLabel}
              label={optionLabel}
              checked={valueMap.includes(optionLabel)}
              disabled={disabled}
              option={option}
              onChange={onChange}
              onMouseEnter={handleMouseEnter}
            />
          );
        })}
      </ul>
      {renderDescription && activeOption && (
        <div className={styles.description}>{renderDescription(activeOption)}</div>
      )}
    </div>
  );
});
