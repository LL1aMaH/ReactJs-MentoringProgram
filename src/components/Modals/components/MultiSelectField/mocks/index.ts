export const mockOptions = [
  'Ann',
  'Jack',
  'Jason',
  'Mary',
  'Carlos',
  'Sara',
  'Matt',
  'Paul',
  'Roberto',
  'Katherine',
  'Robby',
  'Diane',
];
