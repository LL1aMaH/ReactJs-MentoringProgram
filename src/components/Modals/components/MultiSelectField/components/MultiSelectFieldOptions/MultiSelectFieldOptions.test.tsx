import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { MultiSelectFieldOptions } from './MultiSelectFieldOptions';
import { mockOptions } from '../../mocks';

describe('MultiSelectFieldOptions', () => {
  const getComponent = (props = {}) =>
    render(
      <MultiSelectFieldOptions
        valueMap={[]}
        options={mockOptions.slice(0, 3)}
        onChange={jest.fn()}
        {...props}
      />,
    );

  it('should render correctly', () => {
    expect(getComponent().asFragment()).toMatchSnapshot();
  });

  it('should show description on mouse enter', () => {
    const { container, getByText } = getComponent({ renderDescription: () => 'description' });

    fireEvent.mouseEnter(container.querySelector('li') as HTMLElement);

    expect(getByText('description')).not.toBeNull();
  });
});
