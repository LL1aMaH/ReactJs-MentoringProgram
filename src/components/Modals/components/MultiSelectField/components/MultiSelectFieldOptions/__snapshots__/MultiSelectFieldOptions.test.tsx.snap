// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectFieldOptions should render correctly 1`] = `
<DocumentFragment>
  <div
    class="layout"
  >
    <ul
      class="options"
    >
      <li
        class="option"
      >
        <div
          class="container checkbox"
        >
          <input
            class="checkboxField"
            data-testid="checkbox"
            id="Ann"
            type="checkbox"
            value=""
          />
          <span
            class="checkMark"
          />
          <label
            class="label"
            for="Ann"
          >
            Ann
          </label>
        </div>
      </li>
      <li
        class="option"
      >
        <div
          class="container checkbox"
        >
          <input
            class="checkboxField"
            data-testid="checkbox"
            id="Jack"
            type="checkbox"
            value=""
          />
          <span
            class="checkMark"
          />
          <label
            class="label"
            for="Jack"
          >
            Jack
          </label>
        </div>
      </li>
      <li
        class="option"
      >
        <div
          class="container checkbox"
        >
          <input
            class="checkboxField"
            data-testid="checkbox"
            id="Jason"
            type="checkbox"
            value=""
          />
          <span
            class="checkMark"
          />
          <label
            class="label"
            for="Jason"
          >
            Jason
          </label>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
